### PR TITLE
Implement grpc interface server side

### PR DESCRIPTION
**Pull request by zaibon**
Part of #29

### CODE REVIEW

File: server/api/grpc/object_api.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
+package grpc
+
+import (
+	"bytes"
+
+	"golang.org/x/net/context"
```
Comment (delandtj) [inline server/api/grpc/object_api.go]: given go-1.8, why not just "context" ?

Comment (zaibon) [inline server/api/grpc/object_api.go]: cause grpc package doesn't use `context` yet but still the old location

File: server/api/grpc/object_test.go [hunk 2]
```diff
@@ -0,0 +1,222 @@
+package grpc
+
+import (
+	"bytes"
+	"context"
```
Comment (Hamdy) [inline server/api/grpc/object_test.go]: wrong context package no?

Comment (Hamdy) [inline server/api/grpc/object_test.go]: wrong context package no?

File: server/api/grpc/object_api.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+	obj := req.GetObject()
+
+	mgr := manager.NewObjectManager(label, api.db)
+
+	if obj.GetReferenceList() == nil {
+		obj.ReferenceList = []string{}
+	}
+
+	refList := [160][]byte{}
```
Comment (iwanbk) [inline server/api/grpc/object_api.go]: what is 160 here? better to make it constant

File: server/api/grpc/object_api.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+	resp := &pb.GetObjectReply{
+		Object: &pb.Object{
+			Key:           key,
+			Value:         obj.Data,
+			ReferenceList: make([]string, 0),
+		},
+	}
+
+	for i := range obj.ReferenceList {
+		bRef := bytes.Trim(obj.ReferenceList[i][:], "\x00")
```
Comment (iwanbk) [inline server/api/grpc/object_api.go]: what is `\x00` here?

File: server/api/rest/oauth2_itsyouonline_middleware.go [hunk 5]
```diff
@@ -37,10 +37,10 @@ MHYwEAYHKoZIzj0CAQYFK4EEACIDYgAES5X8XrfKdx9gYayFITc89wad4usrk0n2
...
 func NewOauth2itsyouonlineMiddleware(db db.DB) *Oauth2itsyouonlineMiddleware {
 	om := Oauth2itsyouonlineMiddleware{
-		describedBy:"headers",
-		field: "Authorization",
-		url: oauth2ServerUrl,
-		db: db,
+		describedBy: "headers",
+		field:       "Authorization",
+		url:         oauth2ServerUrl,
+		db:          db,
```
Comment (iwanbk) [inline server/api/rest/oauth2_itsyouonline_middleware.go]: someone need to integrate `go-fmt` to his editor

File: server/main.go [hunk 6]
```diff
@@ -79,20 +90,37 @@ func main() {
 		// if err := ensureStoreStat(settings.DB.Dirs.Data, db); err != nil {
 		// 	log.Fatalf("Error checking server stats : %v", err)
 		// }
-
-		server, err := storserver.New(settings.DB.Dirs.Data, settings.DB.Dirs.Meta)
+		db, err := badger.New(settings.DB.Dirs.Data, settings.DB.Dirs.Meta)
```
Comment (iwanbk) [inline server/main.go]: why not creating the `db` in `storserver.New`? By moving  it back here, we undo the work done for #94

File: server/storserver/grpc.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+	lis        net.Listener
+	grpcServer *grpc0.Server
+}
+
+// NewGRPC creates an grpc server with given DB data & meta directory
+func NewGRPC(db db.DB) (StoreServer, error) {
+
+	s := &grpcServer{
+		db:         db,
+		ctx:        context.Background(),
```
Comment (iwanbk) [inline server/storserver/grpc.go]: I don't see the `ctx` being used anywhere.
Is it OK for future use though

File: server/api/grpc/object_api.go (hunk 4 above)
Comment (zaibon) [inline server/api/grpc/object_api.go]: To trim the null bytes at the end of the array. But I just found https://godoc.org/bytes#TrimSpace
That's probable a better way to do it

File: server/api/rest/oauth2_itsyouonline_middleware.go (hunk 5 above)
Comment (zaibon) [inline server/api/rest/oauth2_itsyouonline_middleware.go]: @Hamdy I think that's your code, right ?

File: server/storserver/grpc.go (hunk 7 above)
Comment (zaibon) [inline server/storserver/grpc.go]: I was planning to use it for grafull shutdown, but seems grpc already provide such a mechanism.
If I don't use this context anywhere I'll remove it

File: server/api/rest/oauth2_itsyouonline_middleware.go (hunk 5 above)
Comment (Hamdy) [inline server/api/rest/oauth2_itsyouonline_middleware.go]: @zaibon yes. sorry

**Pull request merged**